File: frontend/types/market.ts
export interface Symbol {
  symbol: string
  base_asset: string
  quote_asset: string
  status: string
  is_spot_trading_allowed: boolean
  is_margin_trading_allowed: boolean
  display_name?: string
}

export interface MarketData {
  symbol: string
  current_price: number
  timestamp: string
  volume_24h: number
  high_24h: number
  low_24h: number
  price_change_24h: number
  price_change_percent_24h: number
  bid_price?: number
  ask_price?: number
  last_trade_price?: number
}

export interface KlineData {
  open_time: number
  open: number
  high: number
  low: number
  close: number
  volume: number
  close_time: number
  quote_volume: number
  trades: number
  taker_buy_base_volume: number
  taker_buy_quote_volume: number
}

export interface OrderBookEntry {
  price: number
  quantity: number
}

export interface OrderBook {
  symbol: string
  bids: OrderBookEntry[]
  asks: OrderBookEntry[]
  last_update_id: number
  timestamp: string
}

export type Timeframe = '1m' | '5m' | '15m' | '30m' | '1h' | '4h' | '1d'

export interface Tab {
  id: string
  symbol: string
  timeframe: Timeframe
  active: boolean
}

File: backend/app/models/market.py
"""
Market data models
"""
from datetime import datetime
from typing import List, Optional, Dict, Any
from pydantic import BaseModel, Field
from enum import Enum


class Timeframe(str, Enum):
    ONE_MINUTE = "1m"
    FIVE_MINUTES = "5m"
    FIFTEEN_MINUTES = "15m"
    THIRTY_MINUTES = "30m"
    ONE_HOUR = "1h"
    FOUR_HOURS = "4h"
    ONE_DAY = "1d"


class PriceData(BaseModel):
    symbol: str
    price: float
    timestamp: datetime
    volume_24h: Optional[float] = None
    price_change_24h: Optional[float] = None
    price_change_percent_24h: Optional[float] = None
    high_24h: Optional[float] = None
    low_24h: Optional[float] = None


class KlineData(BaseModel):
    open_time: int
    open: float
    high: float
    low: float
    close: float
    volume: float
    close_time: int
    quote_volume: float
    trades: int
    taker_buy_base_volume: float
    taker_buy_quote_volume: float


class OrderBookEntry(BaseModel):
    price: float
    quantity: float


class OrderBook(BaseModel):
    symbol: str
    bids: List[OrderBookEntry]
    asks: List[OrderBookEntry]
    last_update_id: int
    timestamp: datetime


class TradeData(BaseModel):
    id: int
    price: float
    quantity: float
    timestamp: datetime
    is_buyer_maker: bool


class MarketData(BaseModel):
    symbol: str
    current_price: float
    timestamp: datetime
    volume_24h: float
    high_24h: float
    low_24h: float
    price_change_24h: float
    price_change_percent_24h: float
    bid_price: Optional[float] = None
    ask_price: Optional[float] = None
    last_trade_price: Optional[float] = None


class WebSocketMessage(BaseModel):
    type: str
    data: Dict[str, Any]
    timestamp: datetime = Field(default_factory=datetime.utcnow)


class PriceUpdate(BaseModel):
    type: str = "price_update"
    symbol: str
    price: float
    volume: Optional[float] = None
    timestamp: datetime = Field(default_factory=datetime.utcnow)


class KlineUpdate(BaseModel):
    type: str = "kline_update"
    symbol: str
    interval: str
    kline: KlineData
    timestamp: datetime = Field(default_factory=datetime.utcnow)


class OrderBookUpdate(BaseModel):
    type: str = "orderbook_update"
    symbol: str
    bids: List[List[float]]  # [[price, quantity], ...]
    asks: List[List[float]]  # [[price, quantity], ...]
    timestamp: datetime = Field(default_factory=datetime.utcnow)


class TradeUpdate(BaseModel):
    type: str = "trade_update"
    symbol: str
    price: float
    quantity: float
    is_buyer_maker: bool
    timestamp: datetime = Field(default_factory=datetime.utcnow)